### PR TITLE
feat: enable optimism hardforks on genesis parse

### LINKED ISSUE #7043: Parse Optimism hardforks from Genesis in `op-reth --chain genesis.json`
### Describe the feature

Currently, `op-reth` doesn't parse Optimism hardforks from Genesis files, with common chains like Base's being [hardcoded](https://github.com/paradigmxyz/reth/blob/91c7dd04b74cfdbff2d0f2df1489850257a1ea56/crates/primitives/src/chain/spec.rs#L256-L286). 
This leads to several execution errors for custom OP chains (like executing a Canyon tx with a Shanghai executor).
Adding this feature will make it easier to run `op-reth` for most OP chains, including devnets!

### Additional context

The relevant code is here:
https://github.com/paradigmxyz/reth/blob/91c7dd04b74cfdbff2d0f2df1489850257a1ea56/crates/node-core/src/args/utils.rs#L66-L102

Ideally, `alloy`'s [Genesis](https://github.com/alloy-rs/alloy/blob/9430040de90ae97fd6677ca830e885ec0f3e7196/crates/genesis/src/lib.rs#L398-L403) would parse `bedrockBlock`, `regolithTime`, etc., for us to update `impl From<Genesis> for ChainSpec` accordingly.
Nevertheless, they have strong reasons not to: https://github.com/alloy-rs/alloy/issues/243.
So, we had to use this current patch to support our custom OP devnet:
```diff
--- a/crates/node-core/src/args/utils.rs
+++ b/crates/node-core/src/args/utils.rs
@@ -1,6 +1,6 @@
 //! Clap parser utilities
 
-use reth_primitives::{fs, AllGenesisFormats, BlockHashOrNumber, ChainSpec, B256};
+use reth_primitives::{fs, AllGenesisFormats, BlockHashOrNumber, ChainSpec, Hardfork, ForkCondition, B256};
 use std::{
     net::{IpAddr, Ipv4Addr, SocketAddr, ToSocketAddrs},
     path::PathBuf,
@@ -95,8 +95,47 @@ pub fn genesis_value_parser(s: &str) -> eyre::Result<Arc<ChainSpec>, eyre::Error
 
             // both serialized Genesis and ChainSpec structs supported
             let genesis: AllGenesisFormats = serde_json::from_str(&raw)?;
+            let mut chain_spec: ChainSpec = genesis.into();
 
-            Arc::new(genesis.into())
+            // RISE TODO: Ideally, alloy's Genesis would parse `bedrockBlock`, `regolithTime`, and `canyonTime`.
+            // They have good reasons not to: https://github.com/alloy-rs/alloy/issues/243.
+            // This hack is therefore needed until we have better OP types in reth, and
+            // `genesis.into()` would set up the custom OP Stack chain's hardforks correctly.
+            #[cfg(feature = "optimism")]
+            {
+                let raw_genesis: serde_json::Value = serde_json::from_str(&raw)?;
+                match raw_genesis["config"]["bedrockBlock"].as_u64() {
+                    Some(block) => {
+                        chain_spec.hardforks.insert(
+                            Hardfork::Bedrock,
+                            ForkCondition::Block(block),
+                        );
+                    }
+                    None => ()
+                }
+                match raw_genesis["config"]["regolithTime"].as_u64() {
+                    Some(time) => {
+                        chain_spec.hardforks.insert(
+                            Hardfork::Regolith,
+                            ForkCondition::Timestamp(time),
+                        );
+                        chain_spec.fork_timestamps = chain_spec.fork_timestamps.regolith(time);
+                    }
+                    None => ()
+                }
+                match raw_genesis["config"]["canyonTime"].as_u64() {
+                    Some(time) => {
+                        chain_spec.hardforks.insert(
+                            Hardfork::Canyon,
+                            ForkCondition::Timestamp(time),
+                        );
+                        chain_spec.fork_timestamps = chain_spec.fork_timestamps.canyon(time);
+                    }
+                    None => ()
+                }
+            }
+
+            Arc::new(chain_spec)
         }
     })
 }
```

Perhaps the actual solution is to build  `alloy-op` or extend the `Genesis` type here in `reth`?
We'd love to contribute PRs once a firm direction is decided :).

### LINKED ISSUE #7702: Add OptimismGenesis fields type
### Describe the feature

Optimism includes additional info in the genesis file.

we don't want to pollute the eth genesis bindings with these, instead we should bundle them in a separate type:

See for example:

https://raw.githubusercontent.com/paradigmxyz/alphanet/331aaea21649e1295d676b73a4ce206bbfbdfe04/crates/testing/resources/etc/genesis.json

This includes OP timestamps + optimism object

This type can be added to node-optimism crate for now

cc @fgimenez 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fgimenez**
Closes #7043, Closes #7702

Modifies the `From<Genesis>`  implementation of `ChainSpec` to take into account the Optimism hardfork info in  `extra_fields`.

### CODE REVIEW

File: crates/primitives/src/chain/spec.rs [hunk 1]
```diff
@@ -1037,6 +1042,21 @@ impl From<Genesis> for ChainSpec {
...
+            #[cfg(feature = "optimism")]
+            (
+                Hardfork::Ecotone,
+                genesis.config.extra_fields.get("ecotoneTime").and_then(|value| value.as_u64()),
+            ),
+            #[cfg(feature = "optimism")]
+            (
+                Hardfork::Canyon,
+                genesis.config.extra_fields.get("canyonTime").and_then(|value| value.as_u64()),
+            ),
```
Comment (mattsse) [inline crates/primitives/src/chain/spec.rs:1059]: this works, but perhaps we could make this reusable by adding a helper struct that contains all the activation info as Optional fields and has a `OptimsismGenesisInfo::extract_from(&Genesis)` or similar

**Review by mattsse — CHANGES_REQUESTED** [review]
nice,
one suggestion for reusability

File: crates/primitives/src/chain/spec.rs (hunk 1 above)
Comment (fgimenez) [reply crates/primitives/src/chain/spec.rs]: sure done here https://github.com/paradigmxyz/reth/pull/7935/commits/1a781c34f29269d050f1195d71c16d89e7d836a7 PTAL

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**